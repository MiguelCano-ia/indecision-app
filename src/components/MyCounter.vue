<template>
  <section>
    <h1 class="text-3xl font-bold underline">Counter: {{ counter }}</h1>
    <h2>Square: {{ square }}</h2>
    <button class="btn mr-2" @click="increment">Increment</button>
    <button class="btn mr-2" @click="decrement">Decrement</button>
    <button class="btn" @click="reset">Reset</button>
  </section>
</template>

<script lang="ts" setup> // lang="ts" es para que el script sea en typescript
  import { ref, computed } from 'vue';

  const props = defineProps<{
    value: number;
  }>();

  const counter = ref(props.value);
  const increment = () => counter.value++; // .value solo es necesario en el setup, en el template no
  const decrement = () => counter.value--;
  const reset = () => counter.value = 0;
  const square = computed(() => counter.value * 2); // Dependencia reactiva de counter

  // No se necesitar un return porque el setup se encarga de eso.
</script>

<style scoped> /* scoped es para que el estilo solo afecte a este componente */

.btn {
  /*Lo que hace el apply es que podamos usar clases de tailwind por separado del html */
  @apply bg-blue-500 text-white font-bold py-2 px-4 rounded;
}

</style>