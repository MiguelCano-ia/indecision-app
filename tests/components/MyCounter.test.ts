import { describe, expect, test } from 'vitest';
import MyCounter from '@/components/MyCounter.vue';
import { mount } from '@vue/test-utils';

describe('<MyCounter />', () => {
  test('Should match snapshot', () => {
    const wrapper = mount(MyCounter, {
      props: {
        value: 0
      }
    });

    expect(wrapper.html()).toMatchSnapshot();
  });

  test('renders the counter value correctly', () => {
    const value = 5;
    const wrapper = mount(MyCounter, {
      props: {
        value: value
      }
    });

    // Imprime lo que contiene h1 del componente console.log((wrapper.find('h1').text()))
    expect(wrapper.find('h1').text()).toContain(value); // Espera que el valor del h1 sea igual al valor que se le paso.
    expect(wrapper.find('[data-testid="square-label"]').text()).toContain(value * value); // Espera que el valor del h1 sea igual a 5.
    // Donde data-testid es un atributo que se le agrega al elemento que se quiere buscar para hacer el test.
  });

  // Pruebas de triggers

  test('increments the counter when +1 button is clicked', async () => {
    const value = 5;
    const wrapper = mount(MyCounter, {
      props: {
        value: value
      },
    });

    const btnIncrement = wrapper.find('button');
    await btnIncrement.trigger('click'); // Se recomienda un await para los triggers

    expect(wrapper.find('h1').text()).toContain(value + 1);
    expect(wrapper.find('[data-testid="square-label"]').text()).toContain((value + 1) * (value + 1));
  });

  test('decrements the counter when -1 is clicked twice', async () => {
    const value = 5;
    const wrapper = mount(MyCounter, {
      props: {
        value: value,
      },
    });

    const [ ,btnDecrement ] = wrapper.findAll('button');
    await btnDecrement.trigger('click');
    await btnDecrement.trigger('click');

    expect(wrapper.find('h1').text()).toContain(value - 2);

  });
});