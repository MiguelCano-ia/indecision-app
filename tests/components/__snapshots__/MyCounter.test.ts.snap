// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`<MyCounter /> > Should match snapshot 1`] = `
"<section data-v-0352012b="">
  <h1 data-v-0352012b="" class="text-3xl font-bold underline">Counter: 0</h1>
  <h2 data-v-0352012b="" data-testid="square-label">Square: 0</h2><button data-v-0352012b="" class="btn mr-2">Increment</button><button data-v-0352012b="" class="btn mr-2">Decrement</button>
</section>"
`;
